src/main/java/com/hierynomus/smbj/share/DiskShare.java

```diff
@@ -76,6 +73,23 @@ public DiskEntry open(String path, Set<AccessMask> accessMask, Set<FileAttribute
         return resolver.handledStates();
     }
 
+    private SMB2CreateResponseContext resolveAndCreateFile(SmbPath path, SMB2ImpersonationLevel impersonationLevel, Set<AccessMask> accessMask, Set<FileAttributes> fileAttributes, Set<SMB2ShareAccess> shareAccess, SMB2CreateDisposition createDisposition, Set<SMB2CreateOptions> createOptions) {
+        DiskShare resolvedShare = this;
+        try {
+            SmbPath target = resolver.resolve(session, path);
+            Session connectedSession = this.session;
+            if (!path.isOnSameHost(target)) {
+                connectedSession = connectedSession.buildNestedSession(target);
```

Comment:
Also not exactly specific to this issue, but I am not sure whether we should not keep track of Connections that get created with buildNestedSession. Session.logoff does not close the connection, so the packet reader thread would be kept alive until the connection times out. (This goes back to #333 and while the fix for #333 fixes session leak, it may not fix connection leak). After the timeout expires and the packet reader thread still hangs around, you'd get EOFException.

Reply:
Actually thinking about it, because the session/connection could've been reused after it was built as a nested session/connection, it's not exactly safe/nice to close them.
Possible solutions:

- ❓  Add a flag to the session/connection that they should not be cached because they're nested. This would work, but it is against the spirit of the specification which says that you should reuse these if possible. Also it would lead to more connections and sessions on the smb server.
- 🚫 Do reference counting and close the session / connection only if it has no child shares. This is almost as dangerous as just closing them, as you don't know whether someone has a hold of the connection/session.
- ✅ The `SmbClient` itself is now `Closeable` to terminate everything. That should solve the problem in its totality.

I'll partly revert the fix for #333 (i.e. the close of the nested session, and I will keep the Closeable SmbClient) because of the above reasoning.